text/0207-context-scoped-attributes.md

```diff
@@ -0,0 +1,245 @@
+As for all APIs, care should be taken to expose it in a way that API-users do not bind themselves to a particular SDK implementation, such as the OpenTelemetry default SDK (for example, by going through a `GlobalOpenTelemetry.GetInstance()` or `GlobalOpenTelemetry.GetInstance().GetContextAttributeWriter()` singleton that returns an interface implementation).
+As this is a write-only API, the API-only implementation should do nothing at all.
+
+### SDK changes
+
+The SDK-implementation could look like this:
+
+```C#
+Context AddContextScopedAttributes(Context context, Attributes attributes) {
+    return context.WithValue(
+        _CTX_KEY_SCOPED_ATTRS,
+        MergeAttributes(context.GetValue(_CTX_KEY_SCOPED_ATTRS), attributes));
+}
+```
+
+The Context-scoped attributes must be available for all telemetry items they are logically associated with.
+For example, the `ReadableSpan` should include the context attributes in the list of span attributes, and they must be included into the list of attributes that the sampler receives.
+This will be an implementation-level change without any changes in the API-surface of the SDK (i.e., it is not necessary to make context-scoped attributes distinguishable from “direct” telemetry attributes).
+
+### Exporter changes (none required)
+
+Exporters should set the attributes on all telemetry items they belong to as if they were set directly on them.
+For example, a span with span attribute `foo=bar` and Context attribute `x=y` would be exported with span attributes `foo=bar, x=y`.
+This should happen by default as the telemetry items will include the context-scoped attributes among their “direct” attributes.
+
+## Trade-offs and mitigations
+
+This OTEP suggests making context-scoped attributes indistinguishable from attributes set directly on the telemetry items.
+This should ease implementation on both backends and OpenTelemetry client libraries, and hopefully also reduces conceptual complexity, but it does makes the feature less powerful.
+No use case is known to the proposer of the OTEP that would require this power though.
+Nevertheless, some implications of making the Context-scoped attributes distinguishable are explored in [Prior art and alternatives](#prior-art-and-alternatives).
+
+This OTEP relies on the [`Context`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/README.md) concept that has been in the spec since 1.0.
+However, at least .NET went 1.0 before the spec and does not include a conforming Context concept.
+For .NET, an async-local variable that holds the Context-scoped attributes seems to be the most useful implementation of this OTEP (similar to how Baggage is implemented there).
+
+## Prior art and alternatives
```

Comment:
I appreciate the extensive research here. I support your proposal in its present, minimal form, with context-scoped attributes indistinguishable from attributes.

When it comes to semantics of attributes, I would like to see us move in a direction where attributes have metadata. One semantic distinction we've discussed elsewhere is that between identifying and non-identifying, descriptive, and ephemeral attributes. 

A number of your alternatives also have to do with compression performance in OTLP, which is not great when you consider all the repetition potentially avoidable. I'd like to keep questions about compression separate.

Reply:
+1 on all 3 paragraphs that @jmacd wrote above.